src/seraphis_mocks/scan_context_async_mock.cpp

```diff
@@ -0,0 +1,797 @@
+    if (top_block_hash == crypto::null_hash && chunk_is_terminal_chunk_out && !chunk_context.block_ids.empty())
+    {
+        m_top_block_hash = chunk_context.block_ids[chunk_context.block_ids.size() - 1];
+        MDEBUG("Used terminal chunk to update top_block_hash " << m_top_block_hash
+            << " (num_blocks_in_chain=" << m_num_blocks_in_chain << ")");
+    }
+
+    if (chunk_is_terminal_chunk_out)
+    {
+        THROW_WALLET_EXCEPTION_IF(m_scan_index.load(std::memory_order_relaxed) < m_end_scan_index,
+            tools::error::wallet_internal_error,
+            "scan index is < m_end_scan_index even though we encountered the terminal chunk");
+
+        THROW_WALLET_EXCEPTION_IF(m_end_scan_index == 0, tools::error::wallet_internal_error,
+            "expected >0 end scan index at terminal chunk");
+
+        THROW_WALLET_EXCEPTION_IF(m_num_blocks_in_chain == 0, tools::error::wallet_internal_error,
+            "expected >0 num blocks in the chain at terminal chunk");
+
+        THROW_WALLET_EXCEPTION_IF(m_top_block_hash == rct::hash2rct(crypto::null_hash),
+            tools::error::wallet_internal_error, "expected top block hash to be set at terminal chunk");
+    }
+}
+//-------------------------------------------------------------------------------------------------------------------
+void AsyncScanContextLegacy::handle_chunk_context(const ChunkRequest &chunk_request,
+    sp::scanning::ChunkContext &chunk_context_out,
+    LegacyUnscannedChunk &unscanned_chunk_out,
+    bool &chunk_is_terminal_chunk_out)
+{
+    // Query daemon for chunk of blocks
+    cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response res = AUTO_VAL_INIT(res);
+    rpc_get_blocks_internal(chunk_request,
+        rpc_get_blocks,
+        m_config.max_get_blocks_attempts,
+        m_config.trusted_daemon,
+        res);
+
+    // Parse the result
+    parse_rpc_get_blocks(chunk_request,
+        res,
+        chunk_context_out,
+        unscanned_chunk_out);
+
+    // Update scanner's known top block height and hash
+    update_chain_state(chunk_context_out,
+        res.current_height,
+        res.top_block_hash,
+        chunk_is_terminal_chunk_out);
+
+    // Check if the chunk was smaller than requested and fill gap if needed
+    fill_gap_if_needed(chunk_is_terminal_chunk_out,
+        chunk_request.requested_chunk_size,
+        chunk_context_out);
+}
+//-------------------------------------------------------------------------------------------------------------------
+async::TaskVariant AsyncScanContextLegacy::chunk_task(const ChunkRequest &chunk_request,
+    std::shared_future<void> &context_stop_flag,
+    std::shared_future<void> &data_stop_flag,
+    std::shared_ptr<std::promise<sp::scanning::ChunkContext>> &chunk_context_ptr_out,
+    std::shared_ptr<std::promise<sp::scanning::ChunkData>> &chunk_data_ptr_out,
+    async::join_token_t &context_join_token_out,
+    async::join_token_t &data_join_token_out)
```

Comment:
```suggestion
    std::shared_future<void> &&context_stop_flag,
    std::shared_future<void> &&data_stop_flag,
    std::promise<sp::scanning::ChunkContext> &&chunk_context_ptr,
    std::promise<sp::scanning::ChunkData> &&chunk_data_ptr,
    async::join_token_t &&context_join_token,
    async::join_token_t &&data_join_token)
```

These are all values that should be moved into the task. Also, not sure what those shared_ptr are for.

Reply:
These changes + moving all values into the task end up triggering `error: static assertion failed: std::function target must be copy-constructible` -- suggestion?

Reply:
Oh uhh, guess leave it as-is then.